kitsune/messages/utils.py

```diff
@@ -18,22 +19,37 @@ def send_message(to, to_group=None, text=None, sender=None):
 
     # Assuming OutboxMessage doesn't necessarily need to be created for every recipient
     if to:
+        # This is users from the To field
+        users = {user for user in to if isinstance(user, User)}
+        # This is groups from the To field
+        groups = [group for group in to if isinstance(group, Group)]
+        # Lets get all the data for groups that we need
+        prefetched_groups = Group.objects.filter(
+            pk__in=[group.pk for group in groups]
+        ).prefetch_related("user_set")
+        # This is who is actually getting a message
+        # - A combo of users from To field and users from groups
+        receivers = users | {user for group in prefetched_groups for user in group.user_set.all()}
+
         outbox_message = OutboxMessage.objects.create(sender=sender, message=text)
-        outbox_message.to.set(to)
-        if to_group:
-            outbox_message.to_group.set(to_group)
-
-    for recipient in to:
-        if isinstance(recipient, User):
-            inbox_message = InboxMessage.objects.create(sender=sender, to=recipient, message=text)
-            # If we had a user, and we made them an inbox message,
-            # we should also add the to_groups to their message as well
-            if to_group:
-                inbox_message.to_group.set(to_group)
-            if Setting.get_for_user(recipient, "email_private_messages"):
-                email_private_message(inbox_message_id=inbox_message.id)
-
-    message_sent.send(sender=InboxMessage, to=to, to_group=to_group, text=text, msg_sender=sender)
+        # Add the users from the To field to the outbox message, not all the users
+        # that are going to receive the message - this way we don't overwhelm the
+        # message UI
+        outbox_message.to.set(users)
+        if groups:
+            # Add the groups from the To field to the message
+            outbox_message.to_group.set(groups)
+
+    for recipient in receivers:
+        inbox_message = InboxMessage.objects.create(sender=sender, to=recipient, message=text)
+        # If we had a user, and we made them an inbox message,
+        # we should also add the groups to their message as well
+        if groups:
+            inbox_message.to_group.set(groups)
+        if Setting.get_for_user(recipient, "email_private_messages"):
```

Comment:
This is a query for every user in the `receivers`, so could be 10's to 100's of queries. Could we make a single bulk query for this before we enter the loop? Maybe something like:

```python
set_of_user_pks_to_email_private_message = set(Setting.objects.filter(user__in=receivers, name="email_private_messages").values_list("user", flat=True))
```

and then, within the loop, check if the recipient's PK is within that set?

Reply:
Good catch! I was wondering if the `value` had to be checked as well.